ddtrace/profiling/collector/_memalloc_tb.cpp

```diff
@@ -61,43 +38,177 @@ traceback_t::init()
+            Py_XDECREF(threading_module);
+            Py_DECREF(get_thread_name_func);
+            threading_module = NULL;
+            get_thread_name_func = NULL;
+            PyErr_SetString(PyExc_RuntimeError, "Failed to get get_thread_native_id from ddtrace.profiling._threading");
+            return false;
+        }
+        // INCREF since we're storing this in a static variable
+        Py_INCREF(get_thread_native_id_func);
+    }
+
     return true;
 }
 
 void
 traceback_t::deinit()
 {
-    // Nothing to clean up - tracebacks are managed by their owners
+    // Clean up DDFrame class reference
+    if (ddframe_class) {
+        Py_DECREF(ddframe_class);
+        ddframe_class = NULL;
+    }
+
+    // Clean up threading function references
+    if (get_thread_name_func) {
+        Py_DECREF(get_thread_name_func);
+        get_thread_name_func = NULL;
+    }
+    if (get_thread_native_id_func) {
+        Py_DECREF(get_thread_native_id_func);
+        get_thread_native_id_func = NULL;
+    }
+    if (threading_module) {
+        Py_DECREF(threading_module);
+        threading_module = NULL;
+    }
 }
 
-frame_t::frame_t(PyFrameObject* pyframe)
+/* Helper function to get thread native_id and name from Python's _threading module
+ * and push threadinfo to the sample.
+ *
+ * NOTE: This is called during traceback construction, which happens during allocation
+ * tracking. We're already inside a reentrancy guard and GC is disabled, so it's safe
+ * to call Python functions here (similar to how we call PyUnicode_AsUTF8String for frames).
+ */
+static void
+push_threadinfo_to_sample(Datadog::Sample& sample, unsigned long thread_id)
 {
-    int lineno_val = PyFrame_GetLineNumber(pyframe);
-    if (lineno_val < 0)
-        lineno_val = 0;
-
-    lineno = (unsigned int)lineno_val;
-
-#ifdef _PY39_AND_LATER
-    PyCodeObject* code = PyFrame_GetCode(pyframe);
-#else
-    PyCodeObject* code = pyframe->f_code;
-#endif
+    // If we don't have a valid thread_id, don't push anything
+    if (thread_id == 0) {
+        return;
+    }
 
-    if (code == NULL) {
-        name = unknown_name;
-        filename = unknown_name;
-    } else {
-        name = code->co_name ? code->co_name : unknown_name;
-        filename = code->co_filename ? code->co_filename : unknown_name;
+    // If threading functions aren't initialized, this is an error condition
+    // (should not happen if initialization succeeded)
+    if (!get_thread_name_func || !get_thread_native_id_func) {
+        // This should not happen - initialization should have failed if functions weren't available
+        // Push empty thread info as fallback
+        int64_t thread_native_id = 0;
+        std::string thread_name("");
+        sample.push_threadinfo(thread_id, thread_native_id, thread_name);
+        return;
     }
 
-    Py_INCREF(name);
-    Py_INCREF(filename);
+    // Get thread native ID from Python
```

Comment:
We can extract the thread tracking logic from echion and make generally available to the profiler, so that we have a convenient way of retrieving thread information within the native layer https://github.com/DataDog/dd-trace-py/blob/53726a1c336664d47b51bb927b76d7337d01dc49/ddtrace/internal/datadog/profiling/stack_v2/src/sampler.cpp#L253